### PR TITLE
Fix stuck hold-actions when modifiers change before key release

### PR DESCRIPTION
**Pull request by MadeByGabe**
`actionHandler:KeyAction` built its release dispatch from the engine-provided `actions` which the engine recomputes from the current modifier state at release time.  

Let's consider a modified bind (e.g. `Shift+n`): If I released `n` while still holding shift, the engine's release dispatch contained `shift+n` and the release handler ran since it was registered to `shift+n`.  
But if I released `Shift` first, only then released `n`, the engine's release dispatch contained only `n` and the release handler never ran since the handler was expecting shift+n.  
The action was stuck in its hold state.  
Same if we changed the modifiers for example by alt-tabbing away, or by simply holding ctrl then releasing shift, etc...

### Work done

Capture the action list each key triggers at press time(keyed by scancode), and use those on release.  
The engine's list is used only as a fallback when nothing was captured.  

#### Test steps
- [ ] Bind for example `Shift+sc_n commandinsert prepend_between`
- [ ] Queue commands whilel holding shift+n. 
- [ ] Release shift, then release n. Commandinsert mode should exit. Without this fix it would stay in insert mode.

I tested all widgets which register hold-style actions, the only issue I found was with `gui_overview_cam_tab_hold.lua` which doesn't work as expected but not because of this fix, it doesn't work correctly without the fix either (neither `gui_overview_keep_camera_position.lua` though that's unrelated). 

Relevant other PR: #7994

**I'm not 100% confident in this fix, I don't know if there are any other edge cases that could cause issues, or that I even understand the engine's behavior well enough to be sure this is a complete/correct fix.** 

### AI / LLM usage statement:

Claude Opus 4.8 validated my reasoning and fix.

### CODE REVIEW

**Comment by WatchTheFort** [comment]
Does this supersede #7994?

**Comment by MadeByGabe** [comment]
Yes @WatchTheFort, almost certainly supersedes it. 

I can't be sure because that PR mentioned Wayland for example and potential focus loss issues which I couldn't reproduce.  
I think that the mentioned focus loss detection issue was just a misdiagnosis though but again I can't be sure.

File: luaui/actions.lua [hunk 1]
```diff
@@ -203,22 +205,43 @@ local function TryAction(actionMap, cmd, optLine, optWords, isRepeat, release, a
...
   end
 
-  for _, bAction in ipairs(actions) do
+  -- Release: prefer the action list captured at press time. 
+  local releaseActions = (scanCode and self.pressedKeyActions[scanCode]) or actions
+  if (scanCode) then
+    self.pressedKeyActions[scanCode] = nil
+  end
+
+  if (not(releaseActions and next(releaseActions))) then return false end
```
Comment (efrec) [inline luaui/actions.lua:237]: style:
```suggestion
  if (not(releaseActions and next(releaseActions))) then
    return false
  end
```

File: luaui/actions.lua [hunk 2]
```diff
@@ -203,22 +205,43 @@ local function TryAction(actionMap, cmd, optLine, optWords, isRepeat, release, a
...
-function actionHandler:KeyAction(press, _, _, isRepeat, _, actions)
-  if (not(actions and next(actions))) then return false end
-
-  local actionSet
+function actionHandler:KeyAction(press, _, _, isRepeat, scanCode, actions)
   if (press) then
-    actionSet = isRepeat and self.keyRepeatActions or self.keyPressActions
-  else
-    actionSet = self.keyReleaseActions
+    if (not(actions and next(actions))) then return false end
```
Comment (efrec) [inline luaui/actions.lua:210]: style:
```suggestion
    if (not(actions and next(actions))) then
	  return false
	end
```

File: luaui/actions.lua [hunk 3]
```diff
@@ -203,22 +205,43 @@ local function TryAction(actionMap, cmd, optLine, optWords, isRepeat, release, a
...
+  for _, bAction in ipairs(releaseActions) do
     local cmd = bAction["command"]
     local extra = bAction["extra"]
     local words = string.split(extra)
 
-    if (TryAction(actionSet, cmd, extra, words, isRepeat, not press, actions)) then
+    if (TryAction(self.keyReleaseActions, cmd, extra, words, false, true, actions)) then
       return true
     end
   end
```
Comment (efrec) [inline luaui/actions.lua:247]: We follow this same pattern twice so it would be good to put in a function.

**Review by efrec — COMMENTED** [review]
Seems good but see notes.

File: luaui/actions.lua (hunk 1 above)
Comment (MadeByGabe) [reply luaui/actions.lua]: Solved by rebase onto master which had the formatting changes already

File: luaui/actions.lua (hunk 2 above)
Comment (MadeByGabe) [reply luaui/actions.lua]: Solved by rebase onto master which had the formatting changes already.

File: luaui/actions.lua (hunk 3 above)
Comment (MadeByGabe) [reply luaui/actions.lua]: Added a TryActionList helper function to replace the two loops.

File: luaui/actions.lua [hunk 4]
```diff
@@ -207,26 +209,50 @@ local function TryAction(actionMap, cmd, optLine, optWords, isRepeat, release, a
...
+		end
 
-	for _, bAction in ipairs(actions) do
-		local cmd = bAction.command
-		local extra = bAction.extra
-		local words = string.split(extra)
+		-- Remember which actions this physical key triggered, so their release can still be dispatched even when the modifier state no longer matches the bind at release time.
+		if scanCode and not isRepeat then
+			self.pressedKeyActions[scanCode] = actions
+		end
```
Comment (efrec) [inline luaui/actions.lua:239]: This returns without setting anything in pressedKeyActions. Any press with an empty action seems like it will leave a dangling modifier, or any interruption between press/release, like holding down a printable char and opening chat.

Probably:

```lua
    local hasActions = actions ~= nil and next(actions) ~= nil

	if press then
		-- Remember which actions this physical key triggered, so their release can still be
		-- dispatched even when the modifier state no longer matches the bind at release time.
		if scanCode and not isRepeat then
			self.pressedKeyActions[scanCode] = hasActions and actions or nil
		end

		if not hasActions then
			return false
		end
```

File: luaui/actions.lua [hunk 5]
```diff
@@ -17,6 +17,8 @@ local actionHandler = {
 	keyRepeatActions = {},
 	keyReleaseActions = {},
 	syncActions = {},
+	-- actions triggered by currently held physical key (keyed by scancode) captured at press time so their release can be dispatched reliably
```
Comment (efrec) [inline luaui/actions.lua]: I think this undersold the change. "Reliably" isn't what's important to know, this makes chained binds release-able at all. Otherwise they are single-key release.

File: luaui/actions.lua [hunk 6]
```diff
@@ -207,26 +209,50 @@ local function TryAction(actionMap, cmd, optLine, optWords, isRepeat, release, a
...
-		local extra = bAction.extra
-		local words = string.split(extra)
+		-- Remember which actions this physical key triggered, so their release can still be dispatched even when the modifier state no longer matches the bind at release time.
+		if scanCode and not isRepeat then
+			self.pressedKeyActions[scanCode] = actions
+		end
 
-		if TryAction(actionSet, cmd, extra, words, isRepeat, not press, actions, key, scanCode) then
+		local actionSet = isRepeat and self.keyRepeatActions or self.keyPressActions
+		if TryActionList(actionSet, actions, isRepeat, false, actions, key, scanCode) then
```
Comment (efrec) [inline luaui/actions.lua:242]: Took me a sec, so this is hardcoded `false` instead of the old `not press`?

File: luaui/actions.lua
Comment (efrec) [inline luaui/actions.lua:181]: idk what happens when an action key is held while the action is removed. I assume that is handled well already and in here.

**Review by efrec — CHANGES_REQUESTED** [review]
I think I need to see more examples of what this is meant to fix since it is speculative but at least with the change below I think I'd understand what is going on.

**Comment by MadeByGabe** [comment]
> I think I need to see more examples of what this is meant to fix since it is speculative but at least with the change below I think I'd understand what is going on.

The shortage of examples might just be a symptom of the problem, I mean hold-release style actions only work reliably on a plain unmodified key and pretty much all of those are bound by default, so widget authors either hack around it or give up and make the feature a toggle.

Two examples I could find are:

```
// this requires the cmd_factoryqmanager.lua default widget to be on and queue preset saved with meta+alt+1 for example
bind shift+sc_b factory_preset_show 

// commandinsert keys
bind Shift+sc_n  commandinsert prepend_between // prepend command into the queue between 2 commands spatially
bind Any+space  commandinsert prepend_queue // prepend command into a separate prepend queue until key released
```

Neither feature cause issues by default because the default hotkeys aren't these.  
I made the above `commandinsert` binds relatively popular because it's very useful, but still that's not the default and the default doesn't cause issues (shift+n does because sometimes player release shift first by accident).  

**I just noticed that the bug happens in even more cases. If a hold-release style action is bound to a plain key such as "n", but then the player holds shift and releases n key, then the same bug happens, key stuck. That's not even an exotic case with chorded binds but with posibbly all binds.**

I also had a third example, the overview camera's hold-release variant was using a hack I believe to avoid this release issue but that widget doesn't work at all now so it doesn't matter either. 

I can find more examples in custom widgets. For example I haven't even tried [this one](https://github.com/beyond-all-reason/BAR-Widgets/blob/38208d3fc1d9234daa750d9bd95e356037a2e1a9/Widgets/community/gui_ping_wheel/gui_ping_wheel.lua#L179-L180) but I can tell that their solution can cause issues unless of course they deal with stuck keys in the update loops which is what many developer had to do.  
They even write "making sure weird lingering display doesn't happen with custom keybind!" which indicates they ran into the same issue.

So just because it's difficult to find easily reproducable bugs doesn't mean there is no issue, see for example the mentioned PR what they had to do to just [fix the same issue with a hack](https://github.com/beyond-all-reason/Beyond-All-Reason/pull/7994/changes).

File: luaui/actions.lua (hunk 4 above)
Comment (MadeByGabe) [reply luaui/actions.lua:239]: Yes sorry you're right, I should have noticed that. To be honest I still don't fully understand everything in this problem space.

File: luaui/actions.lua (hunk 5 above)
Comment (MadeByGabe) [reply luaui/actions.lua]: Yes. Edited the comment to be more precise. I just noticed I made a grammar mistake in that comment but I don't want to make another commit just to fix that.

File: luaui/actions.lua (hunk 6 above)
Comment (MadeByGabe) [reply luaui/actions.lua:242]: Sorry I asked claude to explain this, I can't explain it better:

> Yes, and it's equivalent. On master press and release shared one call site, so it had to pass `not press`. Splitting the branches means `not press` is always `false` inside the press branch, and the release branch always passes `true`. No behaviour change — just no longer recomputing a constant.

File: luaui/actions.lua
Comment (MadeByGabe) [reply luaui/actions.lua:181]: I spent some significant time investigating to give you a good answer to this question and I'm still not confident in anything I could say. I don't think there is an issue with actions removed while key is held or if there is it's a problem that already existed.

File: luaui/actions.lua (hunk 4 above)
Comment (efrec) [reply luaui/actions.lua:239]: Well good news because I think this is the major fix

File: luaui/actions.lua (hunk 6 above)
Comment (efrec) [reply luaui/actions.lua:242]: Yep all good

File: luaui/actions.lua
Comment (efrec) [reply luaui/actions.lua:181]: Tabling this, thanks for looking into it.

**Comment by efrec** [comment]
Very thorough, thank you, I would not be able to review well without this. Added back to my review stack.

My goal isn't to fix the one bug but to make chained bindings releasable, so it's a deeper check. If you want to back off and merge the fix at any point lmk and the process can continue to another code change.

**Comment by MadeByGabe** [comment]
I think it's perfectly fine to wait for the broader fix because the actual bug currently just doesn't affect too many players (it does affect me but it's fine).  

Also this is core functionality, being more careful with it is probably a good idea, even though I've tested my fix in a lot of ways I still can't be confident it doesn't cause problems somewhere.

The only thing I'd ask is that it doesn't sit too long, as a player I'd really like to stop hitting this. But no rush on my account, and thanks for reviewing it properly.

File: luaui/actions.lua [hunk 7]
```diff
@@ -17,6 +17,9 @@ local actionHandler = {
 	keyRepeatActions = {},
 	keyReleaseActions = {},
 	syncActions = {},
+	-- Actions triggered by a currently held physical key, keyed by scancode and captured at press time.
+	-- The engine rebuilds the action list from the live modifier state at release time, so a modified bind (e.g. Shift+n) gets no matching release dispatch once the modifier is let go first. Capturing the press list is makes modified binds releasable.
```
Comment (efrec) [inline luaui/actions.lua]: Worth adding both relevant modes in detail:
```suggestion
	-- Actions triggered by a currently held physical key, keyed by scancode and captured at press time.
	-- The engine rebuilds the action list at release time from a single key only with the live modifier state.
	-- So a modified bind (e.g. Shift+n) gets no matching release dispatch once the modifier is let go first.
	-- Capturing the press list is makes chained or modified binds releasable.
```

**Review by efrec — CHANGES_REQUESTED** [review]
Single-key release and chained-key release both work, though with differences between the two, which is not ideal. For any key bound multiple ways, we would have to register which action, specifically, actually handled the press, in order to release it reliably.

I think that could be done in another change rather than as a follow-on to this PR. This is plenty good enough. I'd approve it.

There are other wanted follow-ups as well:

- Repeat during a held press has an engine timeout, 0.75s, which might cause the repeat handler to fire (repeatedly) until real-release. Needs testing before trying to fix, I had to come up with novel keybinds to test this and don't want a... repeat.
- Chord semantics are currently unwanted, so changes leading to them would be rejected. _I_ want them, though, and they are a small change away.
- The textOwner in e.g. chat can consume a press but allow a release, giving us a release we don't know how to process. Not sure what to do here. This also happens with rmlui, luaui crashes, etc; how can this be made reliable?
- Any number of small engine changes. Unknown.

**Comment by MadeByGabe** [comment]
Thanks, I applied the comment suggestions after fixing my earlier typo.  

I spent a lot of time testing and investigating with the mentioned cases in mind and I've found two things:  

Found a 'repeat' action that requires release: [drawinmap](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/878519b457306d48beff7d5ab6ffff575be3e8a4/luaui/Widgets/gui_chat.lua#L4591)

This had the modifier issue but still has it after this PR and I'm pretty sure because it was made with a [hack to avoid the issue](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/878519b457306d48beff7d5ab6ffff575be3e8a4/luaui/Widgets/gui_chat.lua#L1090-L1102) the PR fixes.  

So press the `` ` `` key to draw on the map, press a modifier, then release the key and often (though for some reason not always) the release action doesn't fire and you keep drawing even after releasing the key.  
I believe this can be easily fixed after this PR is merged though I personally would use a different approach to draw in the first place, but that's a different topic.

---

The second thing I found is an actual though invisible bug this PR would have caused. I found this completely by accident while testing the above and there's some chance there are more cases like this and I have no idea how to even search for them.

So the above widget registers a [drawlabel](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/878519b457306d48beff7d5ab6ffff575be3e8a4/luaui/Widgets/gui_chat.lua#L4590) only 'Press' action which then calls startMapmarkInput() which turns the widget into the textOwner but [widgetHandler:KeyRelease](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/878519b457306d48beff7d5ab6ffff575be3e8a4/luaui/barwidgets.lua#L1993-L1998) has a different branch for textOwner which in effect made a stale entry in the pressedKeyActions table.  

Again this wouldn't have been a problem but could have been in the future.  
I don't really like the solution I committed but I don't know how to do it any other way and I think it's probably fine.

---

I wasn't exactly sure what you meant by "chord semantics" but I tested for example the following `bind sc_q,sc_w commandinsert prepend_between` bind which I would call chained bind. Requires two keys to be pressed after one another and the second key to be held down and this works fine with this PR. I will try to find some more time to perhaps make a widget specifically to test this and other edge cases but currently I think these just work fine.  
Chord would be something like sc_q+sc_w right? I actually didn't even know that was possible to bind but I will try to test that too.

**Comment by efrec** [comment]
I had not meant to include those changes in this PR so regret the delay to merge its main fix. Reviewing, I might back off some of the commits if they're not ready to go.

edit: yay

**Review by efrec — APPROVED** [review]
LGTM
